In [8]:
pip install CBFV

Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 539.2/539.2 kB 2.9 MB/s  0:00:00
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install matminer 

Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
  Using cached matminer-0.10.1-py3-none-any.whl.metadata (4.6 kB)
  Using cached pymongo-4.17.0-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (10 kB)
  Using cached dnspython-2.8.0-py3-none-any.whl.metadata (5.7 kB)
Using cached matminer-0.10.1-py3-none-any.whl (5.3 MB)
Using cached pymongo-4.17.0-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (1.8 MB)
Using cached dnspython-2.8.0-py3-none-any.whl (331 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3/3 [matminer]━━━ 2/3 [matminer]
Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd
df = pd.read_csv("Cleaned_Data.csv")

In [23]:
df.head(10)

,Temperature,Formula,ZT
0,350,Pb0.7Mn0.3Te1Na0.04,0.276253
1,400,Pb0.7Mn0.3Te1Na0.04,0.420050
2,450,Pb0.7Mn0.3Te1Na0.04,0.563897
3,500,Pb0.7Mn0.3Te1Na0.04,0.689989
4,550,Pb0.7Mn0.3Te1Na0.04,0.799702
5,600,Pb0.7Mn0.3Te1Na0.04,0.897287
6,650,Pb0.7Mn0.3Te1Na0.04,0.987101
7,700,Pb0.7Mn0.3Te1Na0.04,1.117313
8,350,Pb0.98Na0.02Te1ISW,0.178594
9,400,Pb0.98Na0.02Te1ISW,0.322591


In [24]:
from pymatgen.core import Composition

df["composition"] = df["Formula"].apply(Composition)

In [25]:
df

,Temperature,Formula,ZT,composition
0,350,Pb0.7Mn0.3Te1Na0.04,0.276253,"(Pb, Mn, Te, Na)"
1,400,Pb0.7Mn0.3Te1Na0.04,0.420050,"(Pb, Mn, Te, Na)"
2,450,Pb0.7Mn0.3Te1Na0.04,0.563897,"(Pb, Mn, Te, Na)"
3,500,Pb0.7Mn0.3Te1Na0.04,0.689989,"(Pb, Mn, Te, Na)"
4,550,Pb0.7Mn0.3Te1Na0.04,0.799702,"(Pb, Mn, Te, Na)"
...,...,...,...,...
1175,550,Na0.015Pb0.955Cd0.03Te1,0.982589,"(Na, Pb, Cd, Te)"
1176,600,Na0.015Pb0.955Cd0.03Te1,1.228149,"(Na, Pb, Cd, Te)"
1177,650,Na0.015Pb0.955Cd0.03Te1,1.438621,"(Na, Pb, Cd, Te)"
1178,700,Na0.015Pb0.955Cd0.03Te1,1.587491,"(Na, Pb, Cd, Te)"


In [26]:
df["composition"]

0       (Pb, Mn, Te, Na)
1       (Pb, Mn, Te, Na)
2       (Pb, Mn, Te, Na)
3       (Pb, Mn, Te, Na)
4       (Pb, Mn, Te, Na)
              ...       
1175    (Na, Pb, Cd, Te)
1176    (Na, Pb, Cd, Te)
1177    (Na, Pb, Cd, Te)
1178    (Na, Pb, Cd, Te)
1179    (Na, Pb, Cd, Te)
Name: composition, Length: 1180, dtype: object

In [27]:
from matminer.featurizers.composition import ElementProperty

featurizer = ElementProperty.from_preset("magpie")

In [28]:
df_features = featurizer.featurize_dataframe(
    df,
    col_id="composition",
    ignore_errors=True
)

ElementProperty:   0%|          | 0/1180 [00:00<?, ?it/s]

In [30]:
df_features

,Temperature,Formula,ZT,composition,MagpieData minimum Number,MagpieData maximum Number,MagpieData range Number,MagpieData mean Number,MagpieData avg_dev Number,MagpieData mode Number,...,MagpieData range GSmagmom,MagpieData mean GSmagmom,MagpieData avg_dev GSmagmom,MagpieData mode GSmagmom,MagpieData minimum SpaceGroupNumber,MagpieData maximum SpaceGroupNumber,MagpieData range SpaceGroupNumber,MagpieData mean SpaceGroupNumber,MagpieData avg_dev SpaceGroupNumber,MagpieData mode SpaceGroupNumber
0,350,Pb0.7Mn0.3Te1Na0.04,0.276253,"(Pb, Mn, Te, Na)",11.0,82.0,71.0,57.519608,16.800269,52.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
1,400,Pb0.7Mn0.3Te1Na0.04,0.420050,"(Pb, Mn, Te, Na)",11.0,82.0,71.0,57.519608,16.800269,52.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
2,450,Pb0.7Mn0.3Te1Na0.04,0.563897,"(Pb, Mn, Te, Na)",11.0,82.0,71.0,57.519608,16.800269,52.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
3,500,Pb0.7Mn0.3Te1Na0.04,0.689989,"(Pb, Mn, Te, Na)",11.0,82.0,71.0,57.519608,16.800269,52.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
4,550,Pb0.7Mn0.3Te1Na0.04,0.799702,"(Pb, Mn, Te, Na)",11.0,82.0,71.0,57.519608,16.800269,52.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1175,550,Na0.015Pb0.955Cd0.03Te1,0.982589,"(Na, Pb, Cd, Te)",11.0,82.0,71.0,65.957500,15.320588,52.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1176,600,Na0.015Pb0.955Cd0.03Te1,1.228149,"(Na, Pb, Cd, Te)",11.0,82.0,71.0,65.957500,15.320588,52.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1177,650,Na0.015Pb0.955Cd0.03Te1,1.438621,"(Na, Pb, Cd, Te)",11.0,82.0,71.0,65.957500,15.320588,52.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1178,700,Na0.015Pb0.955Cd0.03Te1,1.587491,"(Na, Pb, Cd, Te)",11.0,82.0,71.0,65.957500,15.320588,52.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0


* So, here we are getting in total 132 vector features.

In [31]:
df_features.shape

(1180, 136)

In [33]:
df_features.columns

Index(['Temperature', 'Formula', 'ZT', 'composition',
       'MagpieData minimum Number', 'MagpieData maximum Number',
       'MagpieData range Number', 'MagpieData mean Number',
       'MagpieData avg_dev Number', 'MagpieData mode Number',
       ...
       'MagpieData range GSmagmom', 'MagpieData mean GSmagmom',
       'MagpieData avg_dev GSmagmom', 'MagpieData mode GSmagmom',
       'MagpieData minimum SpaceGroupNumber',
       'MagpieData maximum SpaceGroupNumber',
       'MagpieData range SpaceGroupNumber', 'MagpieData mean SpaceGroupNumber',
       'MagpieData avg_dev SpaceGroupNumber',
       'MagpieData mode SpaceGroupNumber'],
      dtype='object', length=136)

In [87]:

X = df_features.drop(columns=["Formula","composition","ZT"]) # Removing no-numeric columns

Y = df_features["ZT"]

In [88]:
X

,Temperature,MagpieData minimum Number,MagpieData maximum Number,MagpieData range Number,MagpieData mean Number,MagpieData avg_dev Number,MagpieData mode Number,MagpieData minimum MendeleevNumber,MagpieData maximum MendeleevNumber,MagpieData range MendeleevNumber,...,MagpieData range GSmagmom,MagpieData mean GSmagmom,MagpieData avg_dev GSmagmom,MagpieData mode GSmagmom,MagpieData minimum SpaceGroupNumber,MagpieData maximum SpaceGroupNumber,MagpieData range SpaceGroupNumber,MagpieData mean SpaceGroupNumber,MagpieData avg_dev SpaceGroupNumber,MagpieData mode SpaceGroupNumber
0,350,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
1,400,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
2,450,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
3,500,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
4,550,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1175,550,11.0,82.0,71.0,65.957500,15.320588,52.0,2.0,90.0,88.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1176,600,11.0,82.0,71.0,65.957500,15.320588,52.0,2.0,90.0,88.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1177,650,11.0,82.0,71.0,65.957500,15.320588,52.0,2.0,90.0,88.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1178,700,11.0,82.0,71.0,65.957500,15.320588,52.0,2.0,90.0,88.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0


In [89]:
Y


0       0.276253
1       0.420050
2       0.563897
3       0.689989
4       0.799702
          ...   
1175    0.982589
1176    1.228149
1177    1.438621
1178    1.587491
1179    1.657323
Name: ZT, Length: 1180, dtype: float64

In [90]:
X.isnull().sum()
# no. of missing values in each columns

Temperature                             0
MagpieData minimum Number              25
MagpieData maximum Number              25
MagpieData range Number                25
MagpieData mean Number                 25
                                       ..
MagpieData maximum SpaceGroupNumber    25
MagpieData range SpaceGroupNumber      25
MagpieData mean SpaceGroupNumber       25
MagpieData avg_dev SpaceGroupNumber    25
MagpieData mode SpaceGroupNumber       25
Length: 133, dtype: int64

In [91]:
missing = X.isnull().sum()

print(missing[missing == 25])

MagpieData minimum Number              25
MagpieData maximum Number              25
MagpieData range Number                25
MagpieData mean Number                 25
MagpieData avg_dev Number              25
                                       ..
MagpieData maximum SpaceGroupNumber    25
MagpieData range SpaceGroupNumber      25
MagpieData mean SpaceGroupNumber       25
MagpieData avg_dev SpaceGroupNumber    25
MagpieData mode SpaceGroupNumber       25
Length: 132, dtype: int64


All the feature columns are having 25 missing values. So, that's why we will use median imputation as missing values are few as compared to total no. of 
rows.

In [93]:
X = X.fillna(X.median())


In [94]:
X.isnull().sum()

Temperature                            0
MagpieData minimum Number              0
MagpieData maximum Number              0
MagpieData range Number                0
MagpieData mean Number                 0
                                      ..
MagpieData maximum SpaceGroupNumber    0
MagpieData range SpaceGroupNumber      0
MagpieData mean SpaceGroupNumber       0
MagpieData avg_dev SpaceGroupNumber    0
MagpieData mode SpaceGroupNumber       0
Length: 133, dtype: int64

In [101]:
Y.shape 

(1180,)

In [100]:
X.shape 


(1180, 133)

In [96]:
# save the feature matrix

feature_df = X.copy()
feature_df["ZT"] = Y

feature_df.to_csv(
    "Magpie_Features.csv",
    index=False
)

In [102]:
df

,Temperature,Formula,ZT,composition
0,350,Pb0.7Mn0.3Te1Na0.04,0.276253,"(Pb, Mn, Te, Na)"
1,400,Pb0.7Mn0.3Te1Na0.04,0.420050,"(Pb, Mn, Te, Na)"
2,450,Pb0.7Mn0.3Te1Na0.04,0.563897,"(Pb, Mn, Te, Na)"
3,500,Pb0.7Mn0.3Te1Na0.04,0.689989,"(Pb, Mn, Te, Na)"
4,550,Pb0.7Mn0.3Te1Na0.04,0.799702,"(Pb, Mn, Te, Na)"
...,...,...,...,...
1175,550,Na0.015Pb0.955Cd0.03Te1,0.982589,"(Na, Pb, Cd, Te)"
1176,600,Na0.015Pb0.955Cd0.03Te1,1.228149,"(Na, Pb, Cd, Te)"
1177,650,Na0.015Pb0.955Cd0.03Te1,1.438621,"(Na, Pb, Cd, Te)"
1178,700,Na0.015Pb0.955Cd0.03Te1,1.587491,"(Na, Pb, Cd, Te)"


In [107]:
df["Formula"].nunique()

139

* so, there are in total 139 unique compounds and we have to split the test and train set by unique compounds.

In [108]:
unique_compounds = df["Formula"].unique()

In [113]:
len(unique_compounds)

139

In [112]:
unique_compounds[:10]

array(['Pb0.7Mn0.3Te1Na0.04', 'Pb0.98Na0.02Te1ISW', 'Pb1Te1Na0.04',
       'Pb1Te0.88S0.12', 'Pb1.00075Zn0.01Te1.01I0.0015',
       'Pb0.97Ce0.03Te1', 'Pb0.999La0.001Te1', 'Pb0.98Na0.02Te1LNSPS',
       'Na0.0125Pb0.9875Se1', 'Pb0.99Na0.01Te1'], dtype=object)

In [ ]:
# now splitting of training and test set using Scikit learn

from sklearn.model_selection import train_test_split

train_compounds, test_compounds = train_test_split(
    unique_compounds,
    test_size=0.20,
    random_state=42
)

In [118]:
print(len(train_compounds))
print(len(test_compounds))

111
28


In [121]:
# Create Masks to get the actual rows for test and train compounds
train_mask = df["Formula"].isin(train_compounds)
test_mask = df["Formula"].isin(test_compounds)
#boolean arrays 

In [124]:
X_train = X[train_mask]
Y_train = Y[train_mask]
X_test = X[test_mask]
Y_test = Y[test_mask]

In [125]:
X_train

,Temperature,MagpieData minimum Number,MagpieData maximum Number,MagpieData range Number,MagpieData mean Number,MagpieData avg_dev Number,MagpieData mode Number,MagpieData minimum MendeleevNumber,MagpieData maximum MendeleevNumber,MagpieData range MendeleevNumber,...,MagpieData range GSmagmom,MagpieData mean GSmagmom,MagpieData avg_dev GSmagmom,MagpieData mode GSmagmom,MagpieData minimum SpaceGroupNumber,MagpieData maximum SpaceGroupNumber,MagpieData range SpaceGroupNumber,MagpieData mean SpaceGroupNumber,MagpieData avg_dev SpaceGroupNumber,MagpieData mode SpaceGroupNumber
0,350,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
1,400,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
2,450,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
3,500,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
4,550,11.0,82.0,71.0,57.519608,16.800269,52.0,2.0,90.0,88.0,...,0.00031,0.000046,0.000078,0.0,152.0,229.0,77.0,188.117647,35.409458,152.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1175,550,11.0,82.0,71.0,65.957500,15.320588,52.0,2.0,90.0,88.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1176,600,11.0,82.0,71.0,65.957500,15.320588,52.0,2.0,90.0,88.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1177,650,11.0,82.0,71.0,65.957500,15.320588,52.0,2.0,90.0,88.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0
1178,700,11.0,82.0,71.0,65.957500,15.320588,52.0,2.0,90.0,88.0,...,0.00000,0.000000,0.000000,0.0,152.0,229.0,77.0,188.065000,36.065000,152.0


In [127]:
X_test

,Temperature,MagpieData minimum Number,MagpieData maximum Number,MagpieData range Number,MagpieData mean Number,MagpieData avg_dev Number,MagpieData mode Number,MagpieData minimum MendeleevNumber,MagpieData maximum MendeleevNumber,MagpieData range MendeleevNumber,...,MagpieData range GSmagmom,MagpieData mean GSmagmom,MagpieData avg_dev GSmagmom,MagpieData mode GSmagmom,MagpieData minimum SpaceGroupNumber,MagpieData maximum SpaceGroupNumber,MagpieData range SpaceGroupNumber,MagpieData mean SpaceGroupNumber,MagpieData avg_dev SpaceGroupNumber,MagpieData mode SpaceGroupNumber
35,350,30.0,82.0,52.0,66.738039,15.10536,52.0,69.0,96.0,27.0,...,0.0,0.0,0.0,0.0,64.0,225.0,161.0,188.267895,36.411893,152.0
36,400,30.0,82.0,52.0,66.738039,15.10536,52.0,69.0,96.0,27.0,...,0.0,0.0,0.0,0.0,64.0,225.0,161.0,188.267895,36.411893,152.0
37,450,30.0,82.0,52.0,66.738039,15.10536,52.0,69.0,96.0,27.0,...,0.0,0.0,0.0,0.0,64.0,225.0,161.0,188.267895,36.411893,152.0
38,500,30.0,82.0,52.0,66.738039,15.10536,52.0,69.0,96.0,27.0,...,0.0,0.0,0.0,0.0,64.0,225.0,161.0,188.267895,36.411893,152.0
39,550,30.0,82.0,52.0,66.738039,15.10536,52.0,69.0,96.0,27.0,...,0.0,0.0,0.0,0.0,64.0,225.0,161.0,188.267895,36.411893,152.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1147,600,34.0,82.0,48.0,62.500000,19.50000,82.0,81.0,90.0,9.0,...,0.0,0.0,0.0,0.0,14.0,225.0,211.0,154.000000,71.000000,225.0
1148,650,34.0,82.0,48.0,62.500000,19.50000,82.0,81.0,90.0,9.0,...,0.0,0.0,0.0,0.0,14.0,225.0,211.0,154.000000,71.000000,225.0
1149,700,34.0,82.0,48.0,62.500000,19.50000,82.0,81.0,90.0,9.0,...,0.0,0.0,0.0,0.0,14.0,225.0,211.0,154.000000,71.000000,225.0
1150,750,34.0,82.0,48.0,62.500000,19.50000,82.0,81.0,90.0,9.0,...,0.0,0.0,0.0,0.0,14.0,225.0,211.0,154.000000,71.000000,225.0


In [128]:
# to check the data leakage 
train_formulas = set(df.loc[train_mask,"Formula"])

test_formulas = set(df.loc[test_mask,"Formula"])

train_formulas.intersection(test_formulas)

set()

So, this indicates the split of train and test is done without any leakage

In [130]:
print(X_train.shape)
print(X_test.shape)

print(Y_train.shape)
print(Y_test.shape)

(948, 133)
(232, 133)
(948,)
(232,)


In [135]:
X_train.to_csv("x_train.csv", index = False)
X_test.to_csv("x_test.csv",index = False)
Y_test.to_csv("y_test.csv",index = False)
Y_train.to_csv("y_train.csv",index = False)

In [136]:
import pickle

with open("compound_split.pkl", "wb") as f:
    pickle.dump((train_compounds, test_compounds), f)